### DISCUSSION TITLE
How to validate route parameters?

### DISCUSSION THREAD

**Discussion by beeyev** [spatie/laravel-data#160] [Q&A]
In Laravel, we can do this trick to be able to validate route parameters during request.
```php
class SomeRequest extends FormRequest
{

    public function validationData(): array
    {
        return array_merge($this->all(), $this->route()->parameters());
    }
}
```

So is there any way to validate route parameters with form data at the same time?

**Comment by francoism90** [comment]
There's a method called `fromRequest`, maybe you could try that one. :)

**Comment by juse-less** [comment]
I was going to propose the same as @francoism90, but I realised it kinda depends on what you'd like to achieve (and where).

The `Data::from*()` methods will only instantiate the `Data` object, by default (no validation).
In order to actually _validate_ the data, you'll need to call [`Data::validate()`](https://github.com/spatie/laravel-data/blob/main/src/Concerns/ValidateableData.php#L21), or, [`Data::validateAndCreate()`](https://github.com/spatie/laravel-data/blob/main/src/Concerns/ValidateableData.php#L46-L49) (which calls `Data::validate()` before returning a new `Data` object, using that validated data).

So, you can either do something like
```php
$data = SomeData::validateAndCreate($request->all());
```
or, as @francoism90 proposed, the [`Data::fromRequest()`](https://spatie.be/docs/laravel-data/v2/as-a-data-transfer-object/request-to-data-object#content-mapping-a-request-onto-a-data-object) method.
```php
class SomeData extends Data
{
    public function __construct(
        // properties
    ) {}

    public static function fromRequest(SomeRequest $request): static
    {
        return self::validateAndCreate($request->all());
    }
}
```

The thing with [`Request::all()`](https://laravel.com/docs/9.x/requests#retrieving-all-input-data) is that it returns [all input data](https://laravel.com/docs/9.x/requests#retrieving-an-input-value), [as well as all query params](https://laravel.com/docs/9.x/requests#retrieving-input-from-the-query-string).

All you need to do is pass the data through the `Data::validate()` method in one way or another.

I believe the problem could be if you have different rules in `SomeRequest`, and `SomeData`, as you'd need to merge the rules (and not only data).
I recall seeing many ways to manipulate rules, but drawing blanks for if this particular case would be possible.

Unrelated to data validation, but posting this from the [Data documentation](https://spatie.be/docs/laravel-data/v2/as-a-data-transfer-object/creating-a-data-object#content-quickly-getting-data-from-models-requests-), just in case it could be of use.
Just under the Model example, there's a couple examples for Laravel Form Requests.

**Comment by karibekov-insigne** [comment]
I am using a solution like this:

```php
<?php declare(strict_types=1);

namespace App\Concerns;

use Illuminate\Http\Request;

trait WithRouteParameters
{
    public static function from(mixed ...$payloads): static
    {
        if ($payloads[0] instanceof Request) {
            $request = $payloads[0];

            $payloads = static::validate(
                array_merge($request->all(), $request->route()?->parameters())
            );
        }

        return parent::from($payloads);
    }
}

```

**Comment by rubenvanassche** [comment]
I think @aidan-casey was working on an attribute which would make it possible to inject route parameters as data properties. Then regular validation would work just like you would expect.
